iModelCore/iModelPlatform/DgnCore/ChangesetTxns.cpp

```diff
@@ -466,6 +466,179 @@ void TxnManager::WriteChangesToFile(BeFileNameCR pathname, DdlChangesCR ddlChang
         m_dgndb.ThrowException("changeset file not created", (int) ChangesetStatus::FileWriteError);
 }
 
+/**
+ * Iterate over local txn and notify changed instance keys and change type.
+ * rootClassFilter is optional but if provided iterator will only iterate over
+ * changes if it to one of root class or any of its derived class.
+ *
+ * Note: For large set of changes this function may consume a lot of none contigious memory.
+ */
+void TxnManager::ForEachLocalChange(std::function<void(ECInstanceKey const&, DbOpcode)> cb, bvector<Utf8String> const& rootClassFilter, bool includeInMemoryChanges) {
+    DbResult rc;
+
+    includeInMemoryChanges = includeInMemoryChanges && HasDataChanges();
+    if (!HasPendingTxns() && !includeInMemoryChanges) {
+        return;
+    }
+
+    // Expand root class filter into all possible derived classes including root class
+    bset<ECClassId> allowedClasses;
+    if (!rootClassFilter.empty()) {
+        bvector<Utf8String> classIdFilter;
+        for(auto& qualifiedName : rootClassFilter) {
+            if (auto classP = m_dgndb.Schemas().FindClass(qualifiedName)) {
+                classIdFilter.push_back(classP->GetId().ToHexStr());
+            } else {
+                m_dgndb.ThrowException(SqlPrintfString("unknown class '%s' provided as filter", qualifiedName.c_str()).GetUtf8CP(), (int) BE_SQLITE_ERROR);
+            }
+        }
+        if (!classIdFilter.empty()) {
+            auto filterStmt = m_dgndb.GetCachedStatement(SqlPrintfString("SELECT [ClassId] FROM [ec_cache_ClassHierarchy] WHERE [BaseClassId] IN (%s) GROUP BY [ClassId]", BeStringUtilities::Join(classIdFilter, ",").c_str()).GetUtf8CP());
+            while(filterStmt->Step() == BE_SQLITE_ROW) {
+                allowedClasses.insert(filterStmt->GetValueId<ECClassId>(0));
+            }
+        }
+    }
+
+    // Group all local txn
+    auto endTxnId = GetCurrentTxnId();
+    auto startTxnId = QueryNextTxnId(TxnId(0));
+    DdlChanges ddlChanges;
+    ChangeGroup dataChangeGroup;
+    for (auto currTxnId = startTxnId; currTxnId < endTxnId; currTxnId = QueryNextTxnId(currTxnId)) {

```

Comment:
`QueryNextTxnId` will return deleted (undone) txns...you probably want to filter those out.

Reply:
Nope, I misread. You use `QueryNextTxnId(0)` to get the *first* txn and iterate until >= curr txn. Deleted txns will always be > curr txn.

Reply:
Correct, that's what I found out.